EditScopeAlgo : Add set membership edits

ericmehl opened the pull request:
This adds the ability to make changes to Set members in an Edit Scope. Within an edit scope, a list of paths can be set to either be included in a set, removed from a set, or neither. The last case passes through Set membership for the paths + Set combination.

A path + Set combination for an Edit Scope can also be queried to determine which of the add / remove / inherit operations are active.

Most of the related code this is meant to support is in `main`, but as this itself isn't an API break, I'm targeting `1.1_maintenance`. I'm happy to rebase and retarget if it makes more sense to keep it on `main`.

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+	addSet->filterPlug()->setInput( addPathFilter->outPlug() );
+	addSet->namePlug()->setInput( spreadsheet->enabledRowNamesPlug() );
+	addSet->modePlug()->setValue( GafferScene::Set::Mode::Add );
+	addSet->setVariablePlug()->setValue( "setMembership:set" );
```
We should use a DeleteContextVariables to prevent this variable from becoming visible outside of the processor. Might also be good to use a ContextMonitor in the unit tests to assert that we don't leak anything.

ericmehl replied:
I think fixed in 4bbbee4beffcbf48d00de396642fb6aacb5eb1a6

Just to check that I understand, it shouldn't currently be leaking a context variable into the scene, so this is a precaution. Is that correct, or is my test not formulated correctly?

I couldn't get the test to fail so looked at `Set` node and saw that the context variable is only present for the filter connected to the Set, not the upstream scene evaluation.

johnhaddon replied:
Sorry Eric, I sent you on a wild goose chase with this one - your analysis of what happens internally in the Set node is correct. We don't need to do any extra context handling of our own, so we can just drop https://github.com/GafferHQ/gaffer/commit/e19f2cbdb28a7ffc4a0341eaeeeaff5000502203.

ericmehl replied:
Cool, dropped before squash down into 83f348a5d5dbd06b4cf0cea7a664918e93ffdad0

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+	SpreadsheetPtr spreadsheet = new Spreadsheet;
+	result->addChild( spreadsheet );
+	spreadsheet->selectorPlug()->setValue( "${setMembership:set}" );
+	spreadsheet->rowsPlug()->addColumn( new StringVectorDataPlug( "add", Plug::Direction::In, new StringVectorData() ) );
```
Would there be any merit in having the column names matching with the SetMembership enum values?

ericmehl replied:
Fixed in 5b8d879faa5a588fd142817501fab1638043f670

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+	);
+
+	Metadata::registerValue( rowsPlug, "spreadsheet:defaultRowVisible", new BoolData( false ) );
+	Metadata::registerValue( rowsPlug->defaultRow(), "spreadsheet:rowNameWidth", new IntData( 300 ) );
```
The other processors probably use this width to accommodate long paths in the row names. Since we're dealing with sets in the row names, we can probably make this slimmer, but should perhaps widen the columns?

ericmehl replied:
Fixed in 7aa071c6c8c5daef2546b55d885b115d5ef17358

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+
+	if( state == EditScopeAlgo::SetMembership::INHERIT )
+	{
+		if( !addPaths.removePaths( paths ) )
```
Since both `add` and `remove` plugs are directly editable by the user as well, I think we'd be better off not assuming that a path is only in one plug or the other. So we should remove from both plugs here...

ericmehl replied:
Fixed in 70b56182b2eeb09a15f6b30c5c427ab16d26d12a

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+	auto rows = processor->getChild<Spreadsheet::RowsPlug>( "edits" );
+
+	auto row = rows->row( set );
+	if( !row )
```
We should also return `INHERIT` if the row exists but is disabled. Which raises the question, what should `setSetMembership()` do when it finds a disabled row? Perhaps it should enable it, but clear out any previous entries from `add` and `remove`?

ericmehl replied:
That sounds reasonable. Fixed in 39c45ac6af6be88fd889900b5785381e31dfb763

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+
+	if( addPaths.match( path ) == PathMatcher::ExactMatch )
+	{
+		return EditScopeAlgo::SetMembership::IN;
```
Since the user can directly edit `add` and `remove` paths via the NodeEditor, it's possible for the same path to be in both. And since the removing node is below the adding node, here we're returning the wrong result for that case. We need to check for removals first, then additions. Or reorder the nodes (I can't think of a good rationale for one order or the other though).

ericmehl replied:
Fixed in 9b7dc9b158938d80412f8f0b046ea7751ee97d17

johnhaddon commented on src/GafferScene/EditScopeAlgo.cpp:
```diff
@@ -854,3 +855,210 @@ const Gaffer::GraphComponent *GafferScene::EditScopeAlgo::attributeEditReadOnlyR
+		return reason;
+	}
+
+	if( state == EditScopeAlgo::SetMembership::IN || state == EditScopeAlgo::SetMembership::INHERIT )
+	{
+		auto addCell = row->cellsPlug()->getChild<Spreadsheet::CellPlug>( g_addSetColumnIndex );
+		if( MetadataAlgo::getReadOnly( addCell ) )
+		{
+			return addCell;
+		}
+
+		auto addPlug = addCell->valuePlug<StringVectorDataPlug>();
+		if( MetadataAlgo::getReadOnly( addPlug ) )
+		{
+			return addPlug;
+		}
+	}
+
+	if( state == EditScopeAlgo::SetMembership::OUT || state == EditScopeAlgo::SetMembership::INHERIT )
+	{
+		auto removeCell = row->cellsPlug()->getChild<Spreadsheet::CellPlug>( g_removeSetColumnIndex );
+		if( MetadataAlgo::getReadOnly( removeCell ) )
+		{
+			return removeCell;
+		}
+
+		auto removePlug = removeCell->valuePlug<StringVectorDataPlug>();
+		if( MetadataAlgo::getReadOnly( removePlug ) )
+		{
+			return removePlug;
+		}
+	}
```
Since `setSetMembership()` needs to modify both `add` and `remove` for all `SetMembership` values, I think we need to drop the outer `if()` blocks here so that we always check both. And maybe we could avoid a bit of repetition by doing `for( columnName : { g_addSetColumnIndex, g_removeSetColumnIndex } )`?

ericmehl replied:
Fixed in c29df46c9baefc0860890b58826d028b85ac4d84

ericmehl commented on include/GafferScene/EditScopeAlgo.h:
```diff
@@ -123,6 +123,22 @@ GAFFERSCENE_API Gaffer::TweakPlug *acquireAttributeEdit( Gaffer::EditScope *scop
+
+enum SetMembership
+{
+	IN,
+	OUT,
+	INHERIT
```
Not sure what spirit in me decided on ALL CAPS, but agreed that it should be exorcised. I like `Inlcuded`, `Excluded` and `Unchanged` too. Another idea for `Unchanged` is `Passthrough`, though then I start thinking `Passthrough`? `PassThrough`?

Let's go with `Unchanged`.

ericmehl replied:
Fixed in 2668ddb9bbc7fe4c3b286735b215c77bc45111dd

johnhaddon replied:
If you're willing to indulge me in some dithering, I'm now thinking `Added` and `Removed` would be better names than `Included` and `Excluded`, because they're a direct match for the terminology on the Set node itself. What do you think? Since I spent all my change-request tokens on the previous rename, I'm happy to make that change myself in a followup PR to save you the bother...

ericmehl replied:
Matching the Set node does seem like an improvement. I have a BOGO going on enum renaming right now, so it's no problem to take care of that. Renamed and squashed into 83f348a5d5dbd06b4cf0cea7a664918e93ffdad0!

ericmehl commented on Changes.md:
```diff
@@ -1,6 +1,11 @@
+API
+---
+
+- EditScopeAlgo : Added methods to modify and query modifications to set members in an Edit Scope.
```
Fixed in rebase, 6ae28e896b0add4cc0aa625a18afe5eb000c8bb2

ericmehl commented on include/GafferScene/EditScopeAlgo.h:
```diff
@@ -123,6 +123,22 @@ GAFFERSCENE_API Gaffer::TweakPlug *acquireAttributeEdit( Gaffer::EditScope *scop
+};
+
+GAFFERSCENE_API void setSetMembership( Gaffer::EditScope *scope, const std::string &set, const IECore::PathMatcher &paths, SetMembership state );
+GAFFERSCENE_API SetMembership getSetMembership( Gaffer::EditScope *scope, const std::string &set, const std::string &path );
```
Fixed in 3a59904ceb3661f98deb47e67abc08743f744de3

ericmehl commented on python/GafferSceneTest/EditScopeAlgoTest.py:
```diff
@@ -1037,5 +1037,110 @@ def testProcessorNames( self ) :
+		for k, v in {"A" : membership.Inherit, "B" : membership.Inherit, "C" : membership.Inherit}.items() :
+			self.assertEqual( GafferScene.EditScopeAlgo.getSetMembership( scope, k, "/group/plane" ), v )
+		for k, v in {"A" : membership.Inherit, "B" : membership.Inherit, "C" : membership.Inherit}.items() :
+			self.assertEqual( GafferScene.EditScopeAlgo.getSetMembership( scope, k, "/group/cube" ), v )
```
Fixed in a1df2594644b2769575fc7b6ce147a7c77f2eda0

ericmehl commented:
I believe I have the comments addressed now. If CI is happy, ready for another look.

johnhaddon commented:
Thanks Eric! I've gone through and resolved all comments except 2, to which I've added additional comments. Sorry, both are due to deficiencies in my original review. If you could drop the DeleteContextVariables commit, I'm happy to merge and do the rename myself if you've had enough of my back and forth...

ericmehl commented:
I dropped the DeleteContextVariables commit, renamed the enum and squashed it all down into 83f348a5d5dbd06b4cf0cea7a664918e93ffdad0. Should be good to go, provided CI agrees.
